core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1434,6 +1447,31 @@ class ReplicaManager(val config: KafkaConfig,
            */
             if (localLog(topicPartition).isEmpty)
               markPartitionOffline(topicPartition)
+            else {
+              val id = topicIds.get(topicPartition.topic())
+              // Ensure we have not received a request from an older protocol
+              if (id != null && !id.equals(Uuid.ZERO_UUID)) {
+                val log = localLog(topicPartition).get
+                // Check if the topic ID is in memory, if not, it must be new to the broker.
+                // If the broker previously wrote it to file, it would be recovered on restart after failure.
+                // If the topic ID is not the default (ZERO_UUID), a topic ID is being used for the given topic.
+                // If the topic ID in the log does not match the one in the request, the broker's topic must be stale.
+                if (!log.topicId.equals(Uuid.ZERO_UUID) && !log.topicId.equals(topicIds.get(topicPartition.topic))) {
+                  stateChangeLogger.warn(s"Topic Id in memory: ${log.topicId.toString} does not" +
+                    s" match the topic Id provided in the request: " +
+                    s"${topicIds.get(topicPartition.topic).toString}.")
+                } else {
+                  // There is not yet a topic ID stored in the log.
+                  // Write the partition metadata file if it is empty.
+                  if (log.partitionMetadataFile.get.isEmpty()) {
+                    log.partitionMetadataFile.get.write(topicIds.get(topicPartition.topic))
+                    log.topicId = topicIds.get(topicPartition.topic)
+                  } else {
+                    stateChangeLogger.warn("Partition metadata file already contains content.")
```

Comment:
why is this a warning?

Reply:
I think that if we reach here, we are in an unexpected state. The partitionMetadata file should have loaded in the topic ID if the file already contained content. I left as a warn rather than an exception.

Reply:
Oops. I think I cleaned up this block and deleted something. There should be a check if log.topicId.equals(id). If so, then the file exists and we shouldn't go in to the block that says "// There is not yet a topic ID stored in the log."

I should also fix the topicIds.get(topicPartition.topic) above and replace with id.

Reply:
I might think about making this code cleaner in general to avoid so many nested if statements

Reply:
So I thought through these cases some more and realized that the metadata file will fail to open if formatted incorrectly. So the only case where there could be data written to the file is if the ID is the zero UUID. So I decided to just fail on reading the file if the zero ID is provided. (We will never write zero ID to file.) The rest of this cleaned up pretty nicely.